File: python/phenomena/particles/mixins/particledata.py
#!/usr/bin/env python

__author__ = "Sebastian Grinschpun"
__license__ = "GPL"
__version__ = "0.1"
__email__ = "sgrinschpun@ifae.es"
__status__ = "Development"

from phenomena.particles.sources import ParticleDataSource

class ParticleData(object):
    '''
    This is a mixin class for the Particle class
    It adds the attributes and methods related to particle data
    Particle data is managed by the ParticleDataSource class
    '''
    @property
    def name(self):
        return self._name

    def _set_name(self, name):
        self._name = name

    @property
    def pdgid(self):
        return self._pdgid

    def _set_pdgid(self, name):
        self._pdgid = ParticleDataSource.getPDGId(name)

    @property
    def mass(self):
        return self._mass

    def _set_mass(self):
        self._mass = ParticleDataSource.getMass(self._name)

    @staticmethod
    def getmass(name):
        return ParticleDataSource.getMass(name)

    @property
    def charge(self):
        return self._charge

    def _set_charge(self):
        self._charge = ParticleDataSource.getCharge(self._name)

    @staticmethod
    def getcharge(name):
        return ParticleDataSource.getCharge(name)

    @property
    def lifetime(self):
        return self._lifetime

    def _set_lifetime(self):
        self._lifetime = ParticleDataSource.getTau(self._name)

    @property
    def type(self):
        return self._type

    def _set_type(self):
        self._type= ParticleDataSource.getType(self._name)

    @property
    def composition(self):
        return self._composition

    def _set_composition(self):
        self._composition = ParticleDataSource.getComposition(self._name)

    @property
    def decay_channels(self):
        return self._decay_channels

    def _set_decayChannels(self):
        self._decay_channels = ParticleDataSource.getDecayChannels(self._name)

    @property
    def baryonnumber(self):
        nq = 0.
        nqbar = 0.
        for q in self._composition:
            if 'bar' in q:
                nqbar += 1
            elif 'bar' not in q:
                nq += 1
        return (nq-nqbar)/3

    @property
    def leptonnumber(self):
        dict = {
        'e-':1,'nu_e':1,
        'e+':-1,'nu_ebar':-1,
        'mu-':1,'nu_mu':1,
        'mu+':-1,'nu_mubar':-1,
        'tau-': 1,'nu_tau':1,
        'tau+':-1,'nu_taubar':-1
        }
        if self.type == 'lepton':
            leptonnumber = dict[self.name]
        else:
            leptonnumber = 0

        return leptonnumber

File: oF/viz1/src/particle/ParticleData.cpp
#include "ParticleData.h"

ParticleData::ParticleData(int _id, int _parentId, string _name, string _type): id(_id), parentId(_parentId), name(_name), type(_type){
}

ParticleData::ParticleData(int _id, int _parentId, string _name, string _type, vector<string> _composition, double _mass, double _charge, vector<string> _decay): id(_id), parentId(_parentId), name(_name), type(_type), composition(_composition), mass(_mass), charge(_charge), decay(_decay){
}

int ParticleData::getId(void){
  return id;
}

int ParticleData::getParentId(void){
  return parentId;
}

string ParticleData::getName(void){
  return name;
}

string ParticleData::getType(void){
  return type;
}

vector<string> ParticleData::getComposition(void){
  return composition;
}

Boolean ParticleData::isFundamental(void){
  return composition.empty();
}

double ParticleData::getMass(void){
  return mass;
}

double ParticleData::getCharge(void){
  return charge;
}

vector<string> ParticleData::getDecay(void){
  return decay;
}

Boolean ParticleData::isStable(void){
  return decay.empty();
}
